definitions/npm/react-redux_v4.x.x/flow_v0.30.x-/test_connect.js

```diff
@@ -0,0 +1,170 @@
+// @flow
+import React from 'react'
+import { connect } from 'react-redux'
+import type { Connector } from 'react-redux'
+
+// copy & paste from redux libedef :(
+type ReduxDispatch<A: { type: $Subtype<string> }> = (action: A) => A;
+
+type Action = { type: 'A' } | { type: 'B' };
+type Dispatch = ReduxDispatch<Action>
+
+type State = {
+  c: number,
+  d: string
+};
+
+type Props1 = {
+  a: number,
+  b: string
+};
+
+const C1 = (props: Props1) => <div>{props.a} {props.b}</div>
+
+type Props2 = {
+  a: number,
+  b: string,
+  dispatch: Dispatch
+};
+
+class C2 extends React.Component<void, Props2, void> {
+  render () {
+    // $ExpectError
+    return <div onClick={this.props.dispatch()}>{this.props.a} {this.props.b}</div>
+  }
+}
+
+class C3 extends React.Component<void, Props2, void> {
+  render () {
+    return <div onClick={this.props.dispatch({ type: 'A' })}>{this.props.a} {this.props.b}</div>
+  }
+}
+
+//
+// connect()
+//
+
+const CC1 = connect()(C1);
+<CC1 a={1} b="s" />;
+// $ExpectError
+<CC1 />; // missing a, b
+// $ExpectError
+<CC1 a={1} />; // missing b
+// $ExpectError
+<CC1 a="s" b="s" />; // wrong a type
+
+const CC3 = connect()(C3);
+<CC3 a={1} b="s" />;
+// $ExpectError
+<CC3 />; // missing a, b
+// $ExpectError
+<CC3 a={1} />; // missing b
+// $ExpectError
+<CC3 a="s" b="s" />; // wrong a type
+
+//
+// connect(null, null, nul, options)
+//
+
+connect(null, null, null, { pure: true })(C1);
+// $ExpectError
+connect(null, null, null, { pure: 1 })(C1); // wrong type
+connect(null, null, null, { withRef: true })(C1);
+// $ExpectError
+connect(null, null, null, { withRef: 1 })(C1); // wrong type
+const CC4 = connect(null, null, null, {})(C1);
+<CC4 a={1} b="s" />;
+// $ExpectError
+<CC4 />; // missing a, b
+
+//
+// connect(mapStateToProps)
+//
+
+type OwnProps1 = { b: string };
```

Comment:
This is the place where I think the new types are loose.
You have to pretty much manually write the props for the wrapped components.

My main effort with the types was to avoid having to write the types for wrappedComponent Props. This is why I think these types are loose. They don't infer as much.

If you're applying the same `connector` to many different props, you'll have to manually type the props for each component twice. Once before it's wrapped, and once after.